### DISCUSSION TITLE
I want better french typography support

### DISCUSSION THREAD

**Discussion by aarongustafson** [WebWeWant/webwewant.fyi#336] [Wants]
Browsers localize automatically quotation marks if a `lang` attribute is provided which is great, but we can do much more to enhance and simplify experience of developers but also authors on the web.

In French, exclamation and question mark both require a non-breaking space before them, on the contrary of English. Often, people on the web use normal space which create orphan sign at the end of a new line. 

Currently, developers aware of this problem solve it by:
- Replacing space characters before ! and ! in text with &nbsp; or \xa0
- Using white-space: nowrap; CSS property and adding an HTML element encompassing the last word and the sign ! or ? of a sentence
- Use no space

Some links to illustrate the issue:

[HTML authoring in French](https://jkorpela.fi/html/french.html)
[Whitespace before some punctuation characters in French: is there a CSS way to avoid lines breaking? - Stack Overflow](https://stackoverflow.com/questions/53851807/whitespace-before-some-punctuation-characters-in-french-is-there-a-css-way-to-a)
[Space Yourself](https://www.smashingmagazine.com/2015/10/space-yourself/)

<hr>

https://webwewant.fyi/wants/5ffb01b21e3e3f6db4eb6404/

**Comment by captainbrosset** [comment]
This seems to me like a fair request.
As the author of the want said, the following html code will produce different results if you change the value of the `lang` attribute:

```html
<!DOCTYPE html>
<html lang="fr">
  <head>
     <meta charset="utf8">
  </head>
  <body>
    <q>this is a quote</q>
  </body>
</html>
```

With `fr`: 
![image](https://user-images.githubusercontent.com/1152698/111276729-f06e0f00-8637-11eb-8f11-de0f636946d7.png)

With `en`:
![image](https://user-images.githubusercontent.com/1152698/111276769-fd8afe00-8637-11eb-94c7-173711031153.png)

I don't know what authoring mechanism could be put in place for this, but something similar to generate correct sentence endings seems useful.

**Comment by javifernandez** [comment]
Hi, 
@bkardell asked me to look at this issue, which I agree it's a fair request indeed. 

As a matter of fact, what is requested might be the expected behavior for any lang using Unicode. The [UAX14](https://www.unicode.org/reports/tr14/) spec states that the exclamation and question marks are defied as [EX](https://www.unicode.org/reports/tr14/#EX) class, which specifically [prohibit](https://www.unicode.org/reports/tr14/#Table1) breaking before these characters. 

> EX | Exclamation/ Interrogation | “!”, “?”, etc. | Prohibit line breaks before

Additionally, the white-space character is defined as [BA](https://www.unicode.org/reports/tr14/#BA) class (break-after), which prohibits breaking before the space.

**Reply by robinwhittleton** [reply]
Does any browser actually implement this though?
